### ISSUE TITLE
[RFC] ONNX Dialect in MLIR

### ISSUE THREAD

**Issue by tjingrant** [onnx/onnx#2499]
Hi ONNX community, we are a team from IBM Research who previously created the onnx-tensorflow project in ONNX; we hope to continue working with the community to develop ONNX dialect in MLIR.

As discussed, we are drawing up a formal proposal for our contribution to enable ONNX in MLIR infrastructure. Specifically, we are interested in contributing and collaborating on the following component/project:

- **ONNX Dialect**: As per discussion, contributing ONNX Dialect to the community is a no-brainer. We only need to work out the technical details for where/how this dialect is going to live. To see an example, please refer to [ONNX AddOp definition](https://gist.github.com/tjingrant/e3398c5461469e6eb716d206a7e337d7#file-1-addop-td) in TableGen.

- **Reference Lowering**: We intend to contribute a naive lowering pipeline connecting high-level ONNX Dialect to lower-level MLIR ones like standard/affine dialects. This pipeline will translate faithfully and naively, ONNX operations to a set of lower-level dialect operations without any subjective opinions attached to the lowered operations (thus, called _reference lowering_). The benefit of having a reference lowering is two folds:
    - **Drive Adoption**: To drive adoption of ONNX Dialect in MLIR community, we must provide concrete ways users can interact with ONNX Dialect, and thereby intuiting it. Reference lowering enables users to (1). Execute the naively lowered MLIR operations, effectively test-driving the ONNX Dialect. (2) Contextualize concretely how ONNX Dialect fit into the broader MLIR infrastructure, by drawing connections with existing dialect and IR components.
    - **Provide Low-Level Interpretation**: In our extensive experience working with ONNX, we know well that ONNX spec can be hard to grasp for engineers working with lower levels of the software stack. And occasionally, the ambiguity inherent in natural language can make interpreting operation descriptions even harder. Reference lowering, on the other hand, provides a rigorous IR definition for ONNX operation semantics and will be a step towards providing a set of low-level, testable, and self-contained interpretation of ONNX standard, with which many engineers will find comfortable working.

    Again, you can see a concrete example lowering [ONNX Dialect](https://gist.github.com/tjingrant/e3398c5461469e6eb716d206a7e337d7#file-2-addop-onnxdialect-mlir) naively to [Affine Dialect](https://gist.github.com/tjingrant/e3398c5461469e6eb716d206a7e337d7#file-3-addop-affinedialect-mlir).

**Comment by liwchang** [comment]
Cool. 
One more point. Shape inference might be added (later).

**Comment by lattner** [comment]
Super exciting - It would be great to see a proper ONNX dialect in upstream MLIR when it is baked out!  This could help provide a nice reference implementation for ONNX

**Comment by liwchang** [comment]
I suddenly notice one thing. ONNX has version support.
I roughly remember there was some discussion about version support in MLIR somewhere.
We might want to check that.

**Comment by antiagainst** [comment]
@liwchang: Yes, I recently sent out a RFC for adding op availability control (versioning, extensions, etc.) to MLIR public mailing list: https://groups.google.com/a/tensorflow.org/forum/#!topic/mlir/UinSwMkz5Fg. Please feel free to share your thoughts there!

**Comment by linkerzhang** [comment]
@tjingrant may you share more thoughts on the motivation of adding this dialect please? 

I'm guessing, is it because that ONNX operator set is so high level that 1) it makes hard for hardware vendors to follow? 2) it makes hard to do compilation based model inference?

I have been thinking that designing low-level operators in ONNX and having high-level operators described/defined as functions.

**Comment by walrusmcd** [comment]
@tjingrant  this is amazing, thanks for sharing !!   Adding to the brainstorming... some of my thoughts for why the onnx dialect is interesting:

- **converters**.  MLIR is part of google's overall strategy to solve converters.    One key scenarios is being able to go from a source tensorFlow training model down to a client inference engine.   All the pain of TOCO (tf-light) and tf2onnx (onnx) could be solved by MLIR/LLVM.    During the training inner loop you immediately know if your op will run on a client (CoreML or WinML/ONNX) since you have targeted that dialect.   
- **IL and interchange**.  if i recall right, MLIR does not have plans for an interchange format.   Nor does it have IL (?)   Their approach is to use LLVM to go from the source training model compiled all the way down to executable machine code as output.  right?   this works if you can compile from source model to hardware every time you need to inference.  Cloud server environments can sometimes do this, but client side environments cannot.    Both cloud and client will be challenged anytime we introduce dynamic hardware diversity.  Either in a container deployed to a virtualized GPU (where you might not know the IHV vendor ahead of time) or on a desktop deployment where you also don't know the GPU IHV vendor ahead of time either.
- **hardware acceleration**.  I think we can use ONNX to help drive IHV hardware support in a open cross vendor way.    The "last mile" of lowering will be after fusion and graph optimization where it will then need a way to lower down to the hardware machine code.    Being able to lower to ONNX ONNX opens up [DirectML](https://docs.microsoft.com/en-us/windows/win32/direct3d12/dml-intro) and the entire windows ecosystem for hardware vendors to build their device drivers (MCDM and WDDM) into the ML ecosystem.   This allows the last mile of lowering to happen on the desktop/cloud deployed client.

Does that match some of your thinking?

On second idea of reference lowering ; I will have to brainstorm more before commenting.   

thanks !!

**Comment by linkerzhang** [comment]
To be clear, I'm not against the idea of having the dialect.

For the 1st point that Paul mentioned above (helping doing the conversion), I don't understand though. TF model can be translated with MLIR to lower-level functions, and ONNX model can be translated into lower level functions. How does that help the conversion please? I mean TF model -> MLIR is good, how MLIR -> ONNX?

**Comment by liwchang** [comment]
@antiagainst: I went through the op availability control proposal, and replies, including 
> We can just write all patterns without caring about required versions/extensions/capabilities and let the legalization framework to reject the ones generating ops unavailable in the target environment. 

I am not sure I should reply here or in google forum, but I think I just start here first.
There is a design choice whether we write all patterns in one op or split to ops in different versions to their own definition. 
I can see the benefits writing all patterns together, say cleaner serialized code and maybe simpler transformation. 
But different from SPIR-V, ONNX has higher chance to be an input IR, instead output IR, if we really want MLIR to be "last mile" for ONNX, instead of just a converter. 
We need to be very careful to avoid ambiguity.
In this way, we might need a global constant to define ONNX's opset in one or multiple mlir files during serialization/deserialization. 
Alternatively, we can split patterns to op definition with different versions, say using postfix, though it looking ugly. 


Basically, two choices.

1.  Define OpA with a union of all patterns, all put hard work in verifier. When serialization/deserialization we need to put a meta about OpSet in files to avoid ambiguity.

2.  Define OpA_v1 and OpA_v2 with their own patterns. 

Any thought or elegant alternative?

**Comment by gramalingam** [comment]
I would split the version-related questions into the following separate questions/issues:

(a) The basic requirement is that we need to "fix" the mathematical-semantics and specs of an MLIR operation (used in the ONNX dialect) forever (as in ONNX). It can never change.

(b) Because of requirement (a), we will need to represent different versions of an op as essentially different ops. What is the recommended MLIR mechanism for dealing with ops that have very closely related semantics/specs? How can we best reuse the infra-structure (like rewrite-rules, transformations, passes) for closely-related ops?

(c) Is it possible in MLIR to have different dialects that share many ops? E.g., would it make sense to model each opset of ONNX as a dialect? This seems hard to do in MLIR, but may be somebody knows better.

**Comment by tjingrant** [comment]
Thanks, @lattner for your encouraging words!

@liwchang Thanks for the heads-up! Regarding shape inference, we are trying to see if we can re-use the existing shape inference infrastructure in ONNX. Obviously we want to stick to existing implementations as much as we can for correctness and efficiency's sake.

@lichang @gramalingam  We also had some thoughts about how to best handle versioning in the ONNX dialect and naive lowering toolchain we've been building and we'd be happy to share thoughts from our perspective/experience working with MLIR and integrate your feedback. I will get back to you on version control a bit later once I iron out some other details including how to ensure important discussions like this get reflected in the code contribution. 

@linkerzhang the biggest motivation is really due to the synergy we see between two great communities we have been working with. In fact what you've described:

> I have been thinking that designing low-level operators in ONNX and having high-level operators described/defined as functions.

Is exactly the "**M**ulti-**L**evel IR" incarnation of MLIR (the others being, for instance, "**M**achine-**L**earning IR").

@walrusmcd I echo your excitement about the possibilities this line of work would enable. I absolutely agree with you in that this work may help with converter efforts and is conducive to reinforcing ONNX's position as the open model exchange format. I'm obviously not as well-versed as you are in the windows ecosystem vocabularies to comment on your third bullet point.

**Comment by KeDengMS** [comment]
This is exciting! Is the plan to create a repo under onnx project, like onnx-mlir (similar to onnx-tensorflow)? Please let me know when your work becomes open source, so we could collaborate.

**Comment by tjingrant** [comment]
@KeDengMS thanks! We are working hard on the logistical side of things and hope to get our codebase on the public Github within 1-2 weeks' timeframe (may roll out some components earlier). onnx-mlir is for sure one of the options! Will definitely keep posting updates here!

**Comment by XVilka** [comment]
Note, that MLIR is now officially a part of the LLVM project: https://mlir.llvm.org/
It should come also with the LLVM 10.0 version, so you can expect much wider adoption after that.

**Comment by bhack** [comment]
For the curious, this is the ONNX thread in MLIR https://groups.google.com/a/tensorflow.org/forum/m/#!topic/mlir/CUL3_8fnJb4

**Comment by bhack** [comment]
@tjingrant I suppose that the thread needs to be moved at https://llvm.discourse.group/c/llvm-project/mlir right?

**Comment by bhack** [comment]
Just for a side info take a look also to https://llvm.discourse.group/t/development-of-high-level-tensor-compute-primitives-dialect-s-and-transformations/388
